### DISCUSSION TITLE
Should LinkML metamodel have a way to specify which slots should be "indexed" in various storage backends, e.g. SQL DDL, mongo, neo4j, etc...

### DISCUSSION THREAD

**Discussion by sierra-moxon** [linkml/linkml#2409] [Ideas]
From a recent discussion with @kheal( see: https://github.com/microbiomedata/nmdc-schema/issues/2155)

The gist of the question is: Should LinkML have a way to specify that a particular slot or set of slots in a class (perhaps using the `unique_keys` meta slot) should result in an index on an attribute/column/property or a compound index (if the storage mechanism supports this—e.g., SQL DDL)?  

Right now, our SQL DDL generator uses the `keys` related metamodel slots to establish FK and PK relationships between tables in that serialization. We get the indexes "for free" in many RDBM implementations.  

For MongoDB, we don't have constraints, but we can specify indexes: https://www.mongodb.com/docs/manual/indexes/. Still, we currently don't serialize a "schema" into MongoDB "collections" because there is no such thing.  But it would be helpful to hint at the code that creates a collection and where to index it. We could do this through the `keys` LinkML mechanism as the SQL DDL generator does to create constraints. 

Thoughts?

**Comment by ialarmedalien** [comment]
I think it'd be useful, yep!

**Comment by aclum** [comment]
Any progress here?

**Comment by nlharris** [comment]
If this is a planned enhancement, we should make an issue for it.

**Comment by aclum** [comment]
Did we end up with any actionable linked issues here?

**Comment by sierra-moxon** [comment]
We have started a linkml-sql working group and that group is in the process of identifying priorities for improvements.  We will add this issue to that group.  But, does anyone have any further thoughts on this besides those we posted to begin with?  
e.g. the action item I see here is that the SQL generator should recognize and use the "keys" metamodel element to generate keys (and by default in PostgreSQL at least, this will also automatically generate indexes -- I am not sure we want to support specific RBMDS dialects beyond sqlite and PostgreSQL).  

In linkml-store: https://github.com/linkml/linkml-store/pull/33 we have the ability to specify indexes, but I know NMDC does not use linkml-store (outside the ontology-loader) to interact with mongodb.  Its not clear to me what LinkML serialization is used to create new collections in MongoDB.  Would you be able to sketch out a technical workflow that you imagine LinkML could help with @aclum or @kheal?  That will help me understand if linkml-store functionality will aid in this process, or if we're looking at a missing feature from a SQL RDBMS implementation.

**Comment by aclum** [comment]
The nmdc example is we wanted a make a combination of two slots the primary key and be able to index on that. The collection in question was https://microbiomedata.github.io/nmdc-schema/functional_annotation_agg/

**Comment by sierra-moxon** [comment]
Thannk @aclum - you might want to check out this modeling paradigm already available for linkml: https://linkml.io/linkml/schemas/constraints.html#unique-key 

in the example above, these four records would pass validation for uniqueness:

symbol: test
atomic_number:   1
neutron_number:  1

symbol:  anothertest
atomic_number:   1
neutron_number: 2

symbol:  thirdtest
atomic_number:  2
neutron_number: 1

symbol:  fourthtest
atomic_number:  1
atomic_number:  3

these would fail:

-- this record would fail because symbol is not unqiue
symbol: test
atomic_number: 5
neutron_number: 6

-- this record would fail because even though symbol is unique, the combination of atomic_number and neutron_number is not unique across the dataset:
symbol: fifthtest
atomic_number:1
neutron_number:2

-- this would fail because both symbol and the combination of atomic_number and neutron_number are not unique:

symbol: thirdtest
atomic_number: 1
neutron_number: 2

I was hoping to ascertain what the technical implementation of the indexing would look like for NMDC so I could suggest how this annotation on a model would translate to the necessary data store (there is no mongo schema generator in LinkML, so if we wanted this index to go into mongodb, we'd need a layer of code on top of the schema -- parsing the nmdc-schema yaml?  or the JSONSchema?) to extract this annotation and make the index.  If the new code used linkml-store, we could use the functionality I pointed out above to create the index.  If we want to write a layer on top of nmdc-schema ourselves, then I guess then it would be up to that layer to parse YAML, or JSONSchema to extract the key annotation.  If we parse the YAML ourselves, of course it is there already, if we want to use JSONSchema instead, then I would need to check how the JSONSchema generator uses this annotation.

Do you know what the technical workflow will be?
